src/metabase/models/serialization/base.clj

```diff
@@ -0,0 +1,171 @@
+  :serdes_type)
+
+(defmethod serdes-dependencies :default [_]
+  [])
+
+(p.types/defprotocol+ ISerializable
+  ;;; Serialization side
+  (serialize-all
+    [this user-or-nil]
+    "Serializes all relevant instances of this model.
+    The return value is a reducible sequence of [file-path entity-map] pairs, possibly empty.
+    There can be a single file for all the entities (eg. settings) or one file per entity.
+
+    The default implementation assumes there will be many files:
+    - It calls (serialize-query this user-or-nil) to get the (lazy, reducible) set of entities to serialize.
+    - For each entity, it calls `(serialize-one e)` and expects a `[file-path {...}]` pair.
+    - These pairs are returned as a reducible stream.
+
+    To serialize as a single file, override this to do something else. (Then `serialize-query` and `serialize-one` are
+    never called and can be stubs.)")
+
+  (serialize-query
+    [this user-or-nil]
+    "Performs the select query, possibly filtered, for all the entities of this type that should be serialized.
+    Returns the result of the db/select-reducible.
+    Defaults to a naive `db/select-reducible` for the entire model, ignoring the optional user argument.
+
+    Only called by `serialize-all`, either the default implementation or (optionally) by a custom one.
+    This exists to be an easy override point, when the default `serialize-all` is good but you need to filter
+    the returned set, eg. by dropping archived entities.")
+
+  (serialize-one
+    [this]
+    "Serializes a single entity into a YAML-friendly map, with its filename.
+    Returns a [file-name {entity map...}] pair.
+
+    Default implementation:
+    - Uses `\"$ENTITY_ID.yaml\"` or `\"$IDENTITY_HASH.yaml\"` as the filename.
+    - Returns the entity as a vanilla map with `:id` and any `:foo_at` fields removed.
+    - Adds the field `:type \"Collection\"` (etc.) to specify the model.
+
+    That suffices for a few simple entities, but most entities will need to override this.
+    They should follow the pattern of dropping timestamps and numeric database IDs, and including the `:type`.")
+
+  ;;; Deserialization side.
+  (deserialize-file
+    [this new-map local]
+    "Given the model, the file contents as an EDN map, and any corresponding entity already existing in the database,
+    this performs whatever steps are necessary to update or insert the incoming entity.
+    Called for each **file**, which usually means for each entity. For single-file output like for settings, this is
+    called once for the whole file.
+    Defaults to [[deserialize-file-plus]] with [[identity]] as the inner transformation.")
+
+  (deserialize-upsert
+    [old-entity new-map]
+    "Given the original entity and the new deserialized map (not an IModel yet), upsert appdb to merge in the updated
+    entity.
+    Defaults to a naive update by primary key, using just the new map's values.")
+
+  (deserialize-insert
```

Comment:
Anything method or function that modifies global state (in this case, the app DB) should end in a `!`. So this should probably be called `deserialize-insert!`

Reply:
Same goes for the other functions with side effects

Reply:
Done, here and elsewhere.